core/src/Cabana_HDF5ParticleOutput.hpp

```diff
@@ -0,0 +1,627 @@
+// HDF5 (XDMF) Particle Field Output.
+//---------------------------------------------------------------------------//
+//! \cond Impl
+// Format traits.
+template <typename T>
+struct HDF5Traits;
+
+template <>
+struct HDF5Traits<short>
+{
+    static hid_t type() { return H5T_NATIVE_SHORT; }
+};
+
+template <>
+struct HDF5Traits<int>
+{
+    static hid_t type() { return H5T_NATIVE_INT; }
+};
+
+template <>
+struct HDF5Traits<float>
+{
+    static hid_t type() { return H5T_NATIVE_FLOAT; }
+};
+
+template <>
+struct HDF5Traits<double>
+{
+    static hid_t type() { return H5T_NATIVE_DOUBLE; }
+};
+
+namespace Impl
+{
+//---------------------------------------------------------------------------//
+// Rank-0 field
+template <class SliceType>
+void writeFields(
+    HDF5Config h5_config, hid_t file_id, int n_local, size_t n_global,
+    hsize_t n_offset, int comm_rank, const char* filename_hdf5,
+    const char* filename_xdmf, const std::string& mesh_name,
+    const SliceType& slice,
+    typename std::enable_if<
+        2 == SliceType::kokkos_view::traits::dimension::rank, int*>::type = 0 )
+{
+    hid_t plist_id;
+    hid_t dset_id;
+    hid_t filespace_id;
+    hid_t memspace_id;
+
+    // HDF5 hyperslab parameters
+    hsize_t offset[1];
+    hsize_t dimsf[1];
+    hsize_t count[1];
+    herr_t status;
+
+    offset[0] = (hsize_t)n_offset;
+    count[0] = (hsize_t)n_local;
+    dimsf[0] = (hsize_t)n_global;
+
+    // Reorder in a contiguous blocked format.
+    Kokkos::View<typename SliceType::value_type*,
+                 typename SliceType::device_type>
+        view( "field", slice.size() );
+    Kokkos::parallel_for(
+        "Cabana::HDF5ParticleOutput::writeFieldRank0",
+        Kokkos::RangePolicy<typename SliceType::execution_space>(
+            0, slice.size() ),
+        KOKKOS_LAMBDA( const int i ) { view( i ) = slice( i ); } );
+
+    // Mirror the field to the host.
+    auto host_view =
+        Kokkos::create_mirror_view_and_copy( Kokkos::HostSpace(), view );
+
+    hid_t type_id = HDF5Traits<typename SliceType::value_type>::type();
+
+    filespace_id = H5Screate_simple( 1, dimsf, NULL );
+    dset_id = H5Dcreate( file_id, slice.label().c_str(), type_id, filespace_id,
+                         H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT );
+
+    H5Sselect_hyperslab( filespace_id, H5S_SELECT_SET, offset, NULL, count,
+                         NULL );
+
+    memspace_id = H5Screate_simple( 1, count, NULL );
+
+    plist_id = H5Pcreate( H5P_DATASET_XFER );
+    // Default IO in HDF5 is independent
+    if ( h5_config.collective )
+        H5Pset_dxpl_mpio( plist_id, H5FD_MPIO_COLLECTIVE );
+
+    H5Dwrite( dset_id, type_id, memspace_id, filespace_id, plist_id,
+              host_view.data() );
+
+    H5Pclose( plist_id );
+    H5Sclose( memspace_id );
+    H5Dclose( dset_id );
+    H5Sclose( filespace_id );
+
+    if ( 0 == comm_rank )
+    {
+        std::string dtype;
```

Comment:
Do you think it's worth helper functions for extracting these?

Reply:
Yes, that would probably be better.

Reply:
I moved this to HDF5Traits.